test/object-store/sync/flx_sync.cpp

```diff
@@ -4791,6 +4791,255 @@ TEST_CASE("flx: pause and resume bootstrapping at query version 0", "[sync][flx]
+
+    // Initialize the realm with some data
+    harness.load_initial_data(fill_person_schema);
+
+    auto config = harness.make_test_file();
+
+    std::mutex hook_mutex;
+    std::function<SyncClientHookAction(std::weak_ptr<SyncSession>, const SyncClientHookData&)> hook_callback;
+
+    config.sync_config->on_sync_client_event_hook = [&hook_callback, &hook_mutex](std::weak_ptr<SyncSession> session,
+                                                                                  const SyncClientHookData& data) {
+        std::lock_guard lock(hook_mutex);
+        if (hook_callback) {
+            return hook_callback(session, data);
+        }
+        return SyncClientHookAction::NoAction;
+    };
+
+    auto realm = Realm::get_shared_realm(config);
+    REQUIRE(!wait_for_download(*realm));
+    REQUIRE(!wait_for_upload(*realm));
+
+
+    // Set up the initial subscription
+    {
+        auto table = realm->read_group().get_table("class_Person");
+        auto new_subs = realm->get_latest_subscription_set().make_mutable_copy();
+        new_subs.insert_or_assign(Query(table));
+        auto subs = new_subs.commit();
+
+        // Wait for subscription update and sync to complete
+        subs.get_state_change_notification(sync::SubscriptionSet::State::Complete).get();
+        REQUIRE(!wait_for_download(*realm));
+        REQUIRE(!wait_for_upload(*realm));
+        wait_for_advance(*realm);
+    }
+    {
+        auto table = realm->read_group().get_table("class_Person");
+        Results results(realm, Query(table));
+        CHECK(results.size() == num_total);
+    }
+
+    // Register connection change callback to verify session was restarted
+    using ConnState = SyncSession::ConnectionState;
+    auto session = realm->sync_session();
+    session->register_connection_change_callback([&machina, &logger](ConnState old_state, ConnState new_state) {
+        logger->trace("Connection state: %1 -> %2", old_state, new_state);
+        machina.transition_with([new_state](TestState curr_state) {
+            switch (new_state) {
+                case ConnState::Disconnected:
+                    return TestState::disconnected;
+                case ConnState::Connecting:
+                    REQUIRE(curr_state == TestState::disconnected);
+                    return TestState::connecting;
+                case ConnState::Connected:
+                    REQUIRE(curr_state == TestState::connecting);
+                    return TestState::connected;
+                default:
+                    FAIL(util::format("Unknown connection state: %1", new_state));
+            }
+        });
+    });
+
+    auto update_perms = [&](std::optional<nlohmann::json> doc_filter, size_t cnt_emps, size_t cnt_mgrs,
+                            size_t cnt_dirs) {
+        using BatchState = sync::DownloadBatchState;
+        std::mutex callback_mutex;
+        BatchState last_state = BatchState::SteadyState;
```

Comment:
Why do we need to track what the last state was? Isn't it valid to finish a bootstrap and immediately start another server-initiated bootstrap?

Reply:
Reworked this function and removed the state transition verification that the last_state var was being used for.